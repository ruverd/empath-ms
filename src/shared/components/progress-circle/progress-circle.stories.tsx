import type { Meta, StoryObj } from "@storybook/react";
import { ProgressCircle } from "./progress-circle";

const meta = {
  title: "Components/ProgressCircle",
  component: ProgressCircle,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
} satisfies Meta<typeof ProgressCircle>;

export default meta;
type Story = StoryObj<typeof ProgressCircle>;

export const Empty: Story = {
  args: {
    percentage: 0,
  },
};

export const Quarter: Story = {
  args: {
    percentage: 25,
  },
};

export const Half: Story = {
  args: {
    percentage: 50,
  },
};

export const ThreeQuarters: Story = {
  args: {
    percentage: 75,
  },
};

export const Full: Story = {
  args: {
    percentage: 100,
  },
};

export const CustomBackground: Story = {
  args: {
    percentage: 60,
  },
  decorators: [
    (Story) => (
      <div className="p-4 bg-shades-violet-1 rounded-lg">
        <Story />
      </div>
    ),
  ],
};

export const MultipleProgress: Story = {
  render: () => (
    <div className="flex gap-4">
      <ProgressCircle percentage={25} />
      <ProgressCircle percentage={50} />
      <ProgressCircle percentage={75} />
    </div>
  ),
};
