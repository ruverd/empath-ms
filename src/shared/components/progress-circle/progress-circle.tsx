import type { ProgressCircleProps } from "./progress-circle.types";

export const ProgressCircle = ({ percentage }: ProgressCircleProps) => {
  const size = 40;
  const strokeWidth = 4;
  const radius = size / 2;
  const normalizedRadius = radius - strokeWidth / 2;
  const circumference = normalizedRadius * 2 * Math.PI;
  const strokeDashoffset = circumference - (percentage / 100) * circumference;

  return (
    <div
      className="relative inline-flex items-center justify-center"
      style={{ width: size, height: size }}
    >
      <svg height={size} width={size} className="rotate-[-90deg]">
        <defs>
          <linearGradient
            id="progressGradient"
            x1="0%"
            y1="0%"
            x2="0%"
            y2="100%"
          >
            <stop offset="0%" stopColor="#5C15A7" />
            <stop offset="100%" stopColor="#0098BD" />
          </linearGradient>
        </defs>
        {/* Background track */}
        <circle
          stroke="#2c004f"
          fill="transparent"
          strokeWidth={strokeWidth}
          r={normalizedRadius}
          cx={radius}
          cy={radius}
        />
        {/* Progress indicator with gradient */}
        <circle
          stroke="url(#progressGradient)"
          fill="transparent"
          strokeWidth={strokeWidth}
          strokeLinecap="round"
          strokeDasharray={circumference + " " + circumference}
          style={{ strokeDashoffset }}
          r={normalizedRadius}
          cx={radius}
          cy={radius}
        />
      </svg>
      {/* Percentage text */}
      <span
        className="absolute text-[#72e9ff] text-[10px] font-bold"
        style={{ transform: "translate(-50%, -50%)", left: "50%", top: "50%" }}
      >
        {percentage}%
      </span>
    </div>
  );
};
