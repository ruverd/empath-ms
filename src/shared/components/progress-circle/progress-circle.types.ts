export interface ProgressCircleProps {
  percentage: number;
}
