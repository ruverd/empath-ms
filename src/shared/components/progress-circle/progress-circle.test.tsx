import { render, screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { ProgressCircle } from "./progress-circle";

describe("ProgressCircle", () => {
  it("should renders with correct percentage text", () => {
    render(<ProgressCircle percentage={75} />);
    expect(screen.getByText("75%")).toBeInTheDocument();
  });

  it("should renders with correct size", () => {
    render(<ProgressCircle percentage={50} />);
    const container = screen.getByText("50%").parentElement;
    expect(container).toHaveStyle({ width: "40px", height: "40px" });
  });

  it("should renders gradient definition", () => {
    render(<ProgressCircle percentage={50} />);
    const gradient = document.querySelector("linearGradient");

    expect(gradient).toHaveAttribute("id", "progressGradient");
    expect(gradient?.querySelector("stop:first-child")).toHaveAttribute(
      "stop-color",
      "#5C15A7"
    );
    expect(gradient?.querySelector("stop:last-child")).toHaveAttribute(
      "stop-color",
      "#0098BD"
    );
  });

  it("should renders with zero percentage", () => {
    render(<ProgressCircle percentage={0} />);
    expect(screen.getByText("0%")).toBeInTheDocument();
  });

  it("should renders with full percentage", () => {
    render(<ProgressCircle percentage={100} />);
    expect(screen.getByText("100%")).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    const { container } = render(<ProgressCircle percentage={50} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
