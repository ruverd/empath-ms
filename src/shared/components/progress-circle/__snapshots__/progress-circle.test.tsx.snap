// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProgressCircle > should match snapshot 1`] = `
<div
  class="relative inline-flex items-center justify-center"
  style="width: 40px; height: 40px;"
>
  <svg
    class="rotate-[-90deg]"
    height="40"
    width="40"
  >
    <defs>
      <lineargradient
        id="progressGradient"
        x1="0%"
        x2="0%"
        y1="0%"
        y2="100%"
      >
        <stop
          offset="0%"
          stop-color="#5C15A7"
        />
        <stop
          offset="100%"
          stop-color="#0098BD"
        />
      </lineargradient>
    </defs>
    <circle
      cx="20"
      cy="20"
      fill="transparent"
      r="18"
      stroke="#2c004f"
      stroke-width="4"
    />
    <circle
      cx="20"
      cy="20"
      fill="transparent"
      r="18"
      stroke="url(#progressGradient)"
      stroke-dasharray="113.09733552923255 113.09733552923255"
      stroke-linecap="round"
      stroke-width="4"
      style="stroke-dashoffset: 56.548667764616276;"
    />
  </svg>
  <span
    class="absolute text-[#72e9ff] text-[10px] font-bold"
    style="transform: translate(-50%, -50%); left: 50%; top: 50%;"
  >
    50
    %
  </span>
</div>
`;
